x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.searchablesnapshots.cache;
+
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.store.AlreadyClosedException;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.common.io.Channels;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+import org.elasticsearch.index.store.BaseSearchableSnapshotIndexInput;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class CacheBufferedIndexInput extends BaseSearchableSnapshotIndexInput {
+
+    private static final int COPY_BUFFER_SIZE = 8192;
+
+    private final CacheDirectory directory;
+    private final long offset;
+    private final long end;
+    private final CacheFileReference cacheFileReference;
+    private final IndexInputStats stats;
+
+    // the following are only mutable so they can be adjusted after cloning
+    private AtomicBoolean closed;
+    private boolean isClone;
+
+    // last read position is kept around in order to detect (non)contiguous reads for stats
+    private long lastReadPosition;
+    // last seek position is kept around in order to detect forward/backward seeks for stats
+    private long lastSeekPosition;
+
+    CacheBufferedIndexInput(CacheDirectory directory, FileInfo fileInfo, IOContext context, IndexInputStats stats) {
+        this("CachedBufferedIndexInput(" + fileInfo.physicalName() + ")", directory, fileInfo, context, stats, 0L, fileInfo.length(),
+            false, null);
+        stats.incrementOpenCount();
+    }
+
+    private CacheBufferedIndexInput(String resourceDesc, CacheDirectory directory, FileInfo fileInfo, IOContext context,
+                                    IndexInputStats stats, long offset, long length, boolean isClone,
+                                    @Nullable CacheFileReference cacheFileReference) {
+        super(resourceDesc, directory.blobContainer(), fileInfo, context);
+        this.directory = directory;
+        this.offset = offset;
+        this.stats = stats;
+        this.end = offset + length;
+        this.closed = new AtomicBoolean(false);
+        this.isClone = isClone;
+        this.cacheFileReference = Objects.requireNonNullElseGet(cacheFileReference,
```

Comment:
Maybe make `CacheFileReference` a `static class` (top-level even) and pass the `directory` to its constructor? That would avoid needing to do this.

Reply:
I must admit I hesitated between the two. I applied your opinion.